lib/communicator_qmp.cpp

```diff
@@ -0,0 +1,357 @@
+      int err_len;                                                                                                     \
+      MPI_Error_string(status, err_string, &err_len);                                                                  \
+      err_string[127] = '\0';                                                                                          \
+      errorQuda("(MPI) %s", err_string);                                                                               \
+    }                                                                                                                  \
+  } while (0)
+
+struct MsgHandle_s {
+  QMP_msgmem_t mem;
+  QMP_msghandle_t handle;
+};
+
+// While we can emulate an all-gather using QMP reductions, this
+// scales horribly as the number of nodes increases, so for
+// performance we just call MPI directly
+#define USE_MPI_GATHER
+
+#ifdef USE_MPI_GATHER
+#include <mpi.h>
+#endif
+
+Communicator::Communicator(int nDim, const int *commDims, QudaCommsMap rank_from_coords, void *map_data,
+                           bool user_set_comm_handle_, void *user_comm)
+{
+  user_set_comm_handle = user_set_comm_handle_;
+
+  int initialized;
+  MPI_CHECK(MPI_Initialized(&initialized));
+
+  if (!initialized) { assert(false); }
+
+  if (user_set_comm_handle) {
+    // The logic here being: previouly all QMP calls are based on QMP_comm_get_default(), and user can
+    // feed in their own MPI_COMM_HANDLE. Now with the following this behavior should remain the same.
+    QMP_COMM_HANDLE = QMP_comm_get_default();
+    MPI_COMM_HANDLE = *((MPI_Comm *)user_comm);
+  } else {
+    QMP_COMM_HANDLE = QMP_comm_get_default();
+    void *my_comm;
+    QMP_get_mpi_comm(QMP_COMM_HANDLE, &my_comm);
+    MPI_COMM_HANDLE = *((MPI_Comm *)my_comm);
+  }
+
+  comm_init(nDim, commDims, rank_from_coords, map_data);
+
+  std::srand(17 * rank + 137);
+}
+
+Communicator::Communicator(Communicator &other, const int *comm_split)
+{
+  user_set_comm_handle = false;
+
+  constexpr int nDim = 4;
+
+  quda::CommKey comm_dims_split;
+
+  quda::CommKey comm_key_split;
+  quda::CommKey comm_color_split;
+
+  for (int d = 0; d < nDim; d++) {
+    assert(other.comm_dim(d) % comm_split[d] == 0);
+    comm_dims_split[d] = other.comm_dim(d) / comm_split[d];
+    comm_key_split[d] = other.comm_coord(d) % comm_dims_split[d];
+    comm_color_split[d] = other.comm_coord(d) / comm_dims_split[d];
+  }
+
+  int key = index(nDim, comm_dims_split.data(), comm_key_split.data());
+  int color = index(nDim, comm_split, comm_color_split.data());
+
+  QMP_comm_split(other.QMP_COMM_HANDLE, color, key, &QMP_COMM_HANDLE);
+  void *my_comm;
+  QMP_get_mpi_comm(QMP_COMM_HANDLE, &my_comm);
+  MPI_COMM_HANDLE = *((MPI_Comm *)my_comm);
+
+  int my_rank_ = QMP_get_node_number();
+
+  QudaCommsMap func = lex_rank_from_coords_dim_t;
+  comm_init(nDim, comm_dims_split.data(), func, comm_dims_split.data());
+
+  std::srand(17 * rank + 137);
```

Comment:
From a robustness standpoint---is this the _only_ place in the code `srand` gets called?

Reply:
Like, the entirety of QUDA. I don't care to deal with this plumbing question during this PR, but we should probably clean this up in one way or another...

Reply:
This is another bug - I have removed this in `communicator_mpi.cpp` but not here. Do we actually need this?